server/src/middleware/rpc/mod.rs

```diff
@@ -0,0 +1,143 @@
+use tower::util::Either;
+
+use jsonrpsee_core::server::MethodResponse;
+use jsonrpsee_types::Request;
+
+/// Similar to `tower::Service` but specific for jsonrpsee and
+/// doesn't requires `&mut self` for performance reasons.
+#[async_trait::async_trait]
+pub trait RpcServiceT<'a> {
+	/// Process a single JSON-RPC call it may be a subscription or regular call.
+	/// In this interface they are treated in the same way but it's possible to
+	/// distinguish those based on the `MethodResponse`.
+	async fn call(&self, request: Request<'a>, transport_label: TransportProtocol) -> MethodResponse;
+}
+
+/// The transport protocol used to send or receive a call or request.
+#[derive(Debug, Copy, Clone)]
+pub enum TransportProtocol {
+	/// HTTP transport.
+	Http,
+	/// WebSocket transport.
+	WebSocket,
+}
+
+impl std::fmt::Display for TransportProtocol {
+	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+		let s = match self {
+			Self::Http => "http",
+			Self::WebSocket => "websocket",
+		};
+
+		write!(f, "{s}")
+	}
+}
+
+/// Context to a JSON-RPC call.
+#[derive(Debug, Clone)]
+pub struct Context {
+	/// Transport protocol.
+	pub transport: TransportProtocol,
+	/// Remote addr.
+	pub remote_addr: SocketAddr,
+	/// Connection id.
+	pub conn_id: usize,
+	/// HTTP Headers.
+	pub headers: HeaderMap,
+	/// URI.
+	pub uri: Uri,
+}
+
+/// Similar to [`tower::ServiceBuilder`] but doesn't
+/// support any middleware implementations.
+#[derive(Debug, Clone)]
+pub struct RpcServiceBuilder<L>(tower::ServiceBuilder<L>);
+
+impl Default for RpcServiceBuilder<Identity> {
+	fn default() -> Self {
+		RpcServiceBuilder(tower::ServiceBuilder::new())
+	}
+}
+
+impl RpcServiceBuilder<Identity> {
+	/// Create a new [`RpcServiceBuilder`].
+	pub fn new() -> Self {
+		Self(tower::ServiceBuilder::new())
+	}
+}
+
+impl<L> RpcServiceBuilder<L> {
+	/// Optionally add a new layer `T` to the [`RpcServiceBuilder`].
+	///
+	/// See the documentation for [`tower::ServiceBuilder::option_layer`] for more details.
+	pub fn option_layer<T>(self, layer: Option<T>) -> RpcServiceBuilder<Stack<Either<T, Identity>, L>> {
+		RpcServiceBuilder(self.0.option_layer(layer))
+	}
+
+	/// Add a new layer `T` to the [`RpcServiceBuilder`].
+	///
+	/// See the documentation for [`tower::ServiceBuilder::layer`] for more details.
+	pub fn layer<T>(self, layer: T) -> RpcServiceBuilder<Stack<T, L>> {
+		RpcServiceBuilder(self.0.layer(layer))
+	}
+
+	/// Add a [`tower::Layer`] built from a function that accepts a service and returns another service.
+	///
+	/// See the documentation for [`tower::ServiceBuilder::layer_fn`] for more details.
+	pub fn layer_fn<F>(self, f: F) -> RpcServiceBuilder<Stack<LayerFn<F>, L>> {
+		RpcServiceBuilder(self.0.layer_fn(f))
+	}
+
+	/// Add a logging layer to [`RpcServiceBuilder`]
+	///
+	/// This logs each request and response for every call.
+	///
+	pub fn rpc_logger(self, max_log_len: u32) -> RpcServiceBuilder<Stack<RpcLoggerLayer, L>> {
```

Comment:
dq: Is this a wrapper over `RpcService.layer(RpcLogger::new(..))` for convenience since that's a very common use-case?